graphql/plugins/palmChatPlugin.js

```diff
@@ -0,0 +1,223 @@
+        const { stream } = parameters;
+    
+        // Define the model's max token length
+        const modelTargetTokenLength = this.getModelMaxTokenLength() * this.getPromptTokenRatio();
+    
+        const palmMessages = this.convertMessagesToPalm(modelPromptMessages || [{ "author": "user", "content": modelPromptText }]);
+        
+        let requestMessages = palmMessages.messages;
+
+        // Check if the token length exceeds the model's max token length
+        if (tokenLength > modelTargetTokenLength) {
+            // Remove older messages until the token length is within the model's limit
+            requestMessages = this.truncateMessagesToTargetLength(requestMessages, modelTargetTokenLength);
+        }
+
+        const context = this.getCompiledContext(text, parameters, prompt.context || palmMessages.context || '');
+        const examples = this.getCompiledExamples(text, parameters, prompt.examples || []);
+
+        const max_tokens = 1024//this.getModelMaxTokenLength() - tokenLength;
+    
+        if (max_tokens < 0) {
+            throw new Error(`Prompt is too long to successfully call the model at ${tokenLength} tokens.  The model will not be called.`);
+        }
+    
+        const requestParameters = {
+            instances: [{
+                context: context,
+                examples: examples,
+                messages: requestMessages,
+            }],
+            parameters: {
+                temperature: this.temperature ?? 0.7,
+                maxOutputTokens: max_tokens,
+                topP: parameters.topP ?? 0.95,
+                topK: parameters.topK ?? 40,
+            }
+        };
+    
+        return requestParameters;
+    }
+
+    // Get the safetyAttributes from the PaLM API Text Completion API response data
+    getSafetyAttributes(data) {
+        const { predictions } = data;
+        if (!predictions || !predictions.length) {
+            return null;
+        }
+
+        // if we got a predictions array back with more than one prediction, return the safetyAttributes of the first prediction
+        if (predictions.length > 1) {
+            return predictions[0].safetyAttributes ?? null;
+        }
+
+        // otherwise, return the safetyAttributes of the content of the first prediction
+        return predictions[0].safetyAttributes ?? null;
+    }
+
+    // Execute the request to the OpenAI Chat API
+    async execute(text, parameters, prompt) {
+        const url = this.requestUrl(text);
+        const requestParameters = this.getRequestParameters(text, parameters, prompt);
+
+        const data = { ...(this.model.params || {}), ...requestParameters };
+        const params = {};
+        const headers = this.model.headers || {};
+        const gcpAuthTokenHelper = this.config.get('gcpAuthTokenHelper');
+        const authToken = await gcpAuthTokenHelper.getAccessToken();
+        headers.Authorization = `Bearer ${authToken}`;
```

Comment:
openai auth headers are integrated at the model level. nothing seems wrong just to consider

Reply:
Yeah in this model they need to be pulled dynamically to make sure they haven't expired (if we do it at config init, they might).